internal/ethapi/tracer.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package ethapi
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/robertkrimen/otto"
+)
+
+// JavascriptTracer provides an implementation of Tracer that evaluates a
+// Javascript function for each VM execution step.
+type JavascriptTracer struct {
+	vm       *otto.Otto 			// Javascript VM instance
+	traceobj *otto.Object 			// User-supplied object to call
+	log      map[string]interface{}	// (Reusable) map for the `log` arg to `step`
+	logvalue otto.Value 			// JS view of `log`
+	err      error 					// Error, if one has occurred
+}
+
+// fakeBig is used to provide an interface to Javascript for 'big.NewInt'
+type fakeBig struct { }
+
+// NewInt creates a new big.Int with the specified int64 value.
+func (fb *fakeBig) NewInt(x int64) *big.Int {
+	return big.NewInt(x)
+}
+
+// NewJavascriptTracer instantiates a new JavascriptTracer instance.
+// code specifies a Javascript snippet, which must evaluate to an expression
+// returning an object with 'step' and 'result' functions.
+func NewJavascriptTracer(code string) (*JavascriptTracer, error) {
+	vm := otto.New()
+
+	// Set up builtins for this environment
+	vm.Set("big", &fakeBig{})
+
+	jstracer, err := vm.Object("(" + code + ")")
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the required functions exist
+	step, err := jstracer.Get("step")
+	if err != nil {
+		return nil, err
+	}
+	if !step.IsFunction() {
+		return nil, fmt.Errorf("Trace object must expose a function step()")
+	}
+
+	result, err := jstracer.Get("result")
+	if err != nil {
+		return nil, err
+	}
+	if !result.IsFunction() {
+		return nil, fmt.Errorf("Trace object must expose a function result()")
+	}
```

Comment:
Wouldn't it be a nice extension that if the `result()` method is missing, the entire trace object is returned? I can imagine that for most trace methods you want everything you collected to be returned, not just part of it, so the result will just be a data dump anyway most of the time.

On a similar note, it might be interesting to allow specifying a single function as the stepper which operates on an implicit object and returns the object at the end.

My random ideas :)